extensions/functions_comparison.yaml

```diff
@@ -57,6 +57,24 @@ scalar_functions:
           - value: any1
           - value: any1
         return: BOOLEAN
+  -
+    name: "between"
```

Comment:
@ianmcook, do you have any thoughts on this?  I believe your reasoning was that you think engines can execute this more efficiently as a single function as opposed to doing the and of lte and gte?

FWIW postgresql, oracle, and sql server all also have this `between` functionality

Reply:
I believe all the SQL systems treat this as sugared syntax, not a discrete function. This is definitely how tools I've worked with have handled it.

Reply:
Does substrait aim to be an engine or a programming language for users?

If an engine, I would not support `between` as a primitive, because there seems to be a cost without much benefit  to increasing the size of the language. Last time I checked, Oracle's engine does not support `between`: it expands `x between y and z` to `x >= y and x <= z`.

For the important case where `y` and `z` are literals, I would recommend using something like Calcite's Sarg concept. A Sarg is a range set literal, very similar to the structures used by many DBMS for range scans, and very amenable to optimization.

With Sargs, `x between 3 and 5` would translate to `x memberOf Sarg '{[3, 5]}'`. The following conditions can all be represented in a single Sarg:
* `x > 10`
* `x = 5 or x between 10 and 20 or x is null`
* `x not between 3 and 6`
* `x in (1, 3, 5, 10, 20)`
* `x is null`
* `x between symmetric 3 and 5`
* `x > 3 and x < 5`

Reply:
It seems like for now we should stick with simplicity and avoid adding a function because of an optimization.

@jacques-n Can we use this notion of sarg to capture some of the literal range use cases?

Reply:
When deciding whether or not to define a not-strictly-necessary function in the Substrait YAML, it seems to me that we should follow the most commonly used SQL dialects. If instead we want to use more restrictive criteria, then we should define what those criteria are more clearly. For example, why should we allow `cos()` to be defined? Producers can convert `cos(x)` to `sin(subtract(divide(3.1415926535897932384626433832795028841971693993751058209749445923078164062, 2), x))` so do we really need it? That question does not seem fundamentally different than this question about whether producers should have to convert `between(x, y, z)` to `and(gte(x, y), lte(x, z))`.

Reply:
I agree that it makes sense to follow SQL dialects. What's sticky in this situation is that between is not a function in SQL. It's a syntactic construct that represents gte + lte. 

I'm fine (+0.5) including it for the same reason we include the orlist expression. We generally don't want to have to make people jump through hoops on consumption to recompose something that was in a more optimal state prior to substrait plan production. (IOW, we could recompose a orlist or between using heuristic tree traversal but why make everyone do that.)

Reply:
> For example, why should we allow `cos()` to be defined? Producers can convert `cos(x)` to `sin(subtract(divide(3.1415926535897932384626433832795028841971693993751058209749445923078164062, 2), x))` so do we really need it? That question does not seem fundamentally different than this question about whether producers should have to convert `between(x, y, z)` to `and(gte(x, y), lte(x, z))`.

I've been thinking about that one too. It'd be so much easier if we could just say "primitives only, period," but this is a perfect counterexample for it. A less obvious one is that you could define `sin` sufficiently accurately with only addition, subtraction, multiplication, and maybe a 1/x operation.

So I'm inclined to agree with what @ianmcook said, but with a side note that I don't think we should follow SQL logic when it's obviously archaic or just weird, can be defined better using features that exist in Substrait but not in SQL, or if it doesn't mesh well with other things in Substrait (such as nested types).